marshal/admin.go

```diff
@@ -0,0 +1,242 @@
+	}
+}
+
+// setKafkaOffsets sets the Kafka offset and the offset within marshal for the given partition,
+// while heartbeating continuously.
+func (a *consumerGroupAdmin) setKafkaOffsets(topic string, partID int, newOffset int64) bool {
+	log.Infof("[%s:%d] Admin attempting to set offset %d for partition", topic, partID, newOffset)
+	stopHeartbeat := make(chan struct{}, 1)
+	defer func() {
+		close(stopHeartbeat)
+	}()
+
+	// Get current offsets, which we will try to keep heartbeating.
+	offsets, err := a.marshaler.GetPartitionOffsets(topic, partID)
+	if err != nil {
+		log.Errorf("[%s:%d] Admin couldn't get current offsets to heartbeat.", topic, partID)
+		return false
+	}
+
+	// Next, try to claim the partition.
+	if !a.marshaler.ClaimPartition(topic, partID) {
+		log.Errorf("[%s:%d] Admin couldn't claim partition to set Kafka offset",
+			topic, partID)
+		return false
+	}
+	atomic.AddInt32(a.steps, 1)
+
+	// Continuously heartbeat the last offsets while resetting them.
+	go a.heartbeatLoop(topic, partID, offsets.Current, stopHeartbeat)
+
+	// Stop the heartbeat loop with the old offset and heartbeat the new offset,
+	// which will commit the offsets to Kafka.
+	stopHeartbeat <- struct{}{}
+	if err := a.marshaler.Heartbeat(topic, partID, newOffset); err != nil {
+		log.Errorf("[%s:%d] Admin failed to heartbeat new offset %d", topic, partID, newOffset)
+		return false
+	}
+	atomic.AddInt32(a.steps, 1)
+
+	// Send a ReleasePartition with the new offset.
+	if err := a.marshaler.ReleasePartition(topic, partID, newOffset); err != nil {
+		log.Errorf("[%s:%d] Admin failed to release partition with new offset %d: %s",
+			topic, partID, newOffset, err)
+		return false
+	}
+	atomic.AddInt32(a.steps, 1)
+
+	return true
+}
+
+// waitForConsumerReleaseClaim waits for the paused consumer to release its claim.
+func (a *consumerGroupAdmin) waitForConsumerReleaseClaim(topic string, partID int) bool {
+	partitionClaim := a.marshaler.GetLastPartitionClaim(topic, partID)
+
+	start := time.Now()
+	for partitionClaim.isClaimed(time.Now().Unix()) {
+		if time.Since(start) > consumerReleaseClaimWaitTime {
+			log.Errorf("[%s:%d] Admin timed out waiting for consumer to release its claim",
+				topic, partID)
+			return false
+		}
+
+		if partitionClaim.GroupID != a.groupID {
+			log.Errorf("Trying to reset offset of a partition not being read by our " +
+				"consumer group. Admin will not reset this partition.")
+			return false
+		}
+
+		time.Sleep(time.Duration(consumerReleaseClaimWaitSleep))
+		partitionClaim = a.marshaler.GetLastPartitionClaim(topic, partID)
+	}
+
+	return true
+}
+
+// ConstructReleaseGroupMessage constructs a ReleaseGroup message to write to the marshal topic.
+func (a *consumerGroupAdmin) ConstructReleaseGroupMessage(topic string, partID int) *msgReleaseGroup {
+	now := time.Now()
+	base := &msgBase{
+		Time:       int(now.Unix()),
+		InstanceID: newInstanceID(),
+		ClientID:   a.clientID,
+		GroupID:    a.groupID,
+		Topic:      topic,
```

Comment:
Talked f2f, realized zorkian meant just zeroing the fields and not the actual topic, partition we produced to.